File: gradient_descent/gradient_descent.py
import sympy as sp
from sympy.utilities.lambdify import lambdify, implemented_function
import numpy as np

# Set up SymPy for symbolic differentiation
x = sp.Symbol('x')
y = (x+5)**2
yprime = y.diff(x)
print("First Derivative: " + str(yprime))

# Convert the first derivative of the input function to a lambda
yprime = lambdify(x, yprime)

# Instantiate gradient descent inputs
cur_x = 10
rate = 0.001
precision = 0.000000000000001
prev_step_size = 1
n = 0 # Current iteration
max_n = 25000 # max iterations

# Loop until precision meets threshold
while prev_step_size > precision and n < max_n:
    prev_x = cur_x
    cur_x = cur_x - rate * yprime(prev_x)
    prev_step_size = abs(cur_x - prev_x)
    n += 1

    # print("Iteration:",n,"\nX:",cur_x)
    
print("The local minimum occurs at", cur_x)

File: gradient_descent/gradientDescent/gradientDescent.go
package main

import (
	"fmt"
	"math"
)

func main() {
	var cur_x float64 = -10.00
	const precision float64 = 0.0000000000001
	var n int = 0
	const max_n int = 25000
	const rate float64 = 0.001
	// var ydiff = (2 * cur_x) + 10

	for prevStepSize := 1.00; prevStepSize > precision && n < max_n; n += 1 {
		prev_x := cur_x
		cur_x = cur_x - rate*((2*cur_x)+10)
		prevStepSize = math.Abs(cur_x - prev_x)
		if n%10 == 0 {
			fmt.Println("Iteration:", n, "\nX:", cur_x)
		}
	}
	fmt.Println("The local minimum occurs at", cur_x)
}
